src/main/java/org/openrewrite/java/template/RefasterTemplateProcessor.java

```diff
@@ -280,17 +281,25 @@ public void visitClassDef(JCTree.JCClassDecl classDecl) {
                                 .map(Map.Entry::getValue)
                                 .flatMap(Set::stream)
                                 .collect(Collectors.toSet());
+                        Set<String> usedTypes = new LinkedHashSet<>();
                         for (String import_ : imports.values().stream().flatMap(Set::stream).collect(Collectors.toSet())) {
                             if (import_.startsWith("java.lang.")) {
                                 continue;
                             }
                             if (beforeImports.contains(import_) && afterImports.contains(import_)) {
+                                usedTypes.add(import_);
                             } else if (beforeImports.contains(import_)) {
                                 recipe.append("                    maybeRemoveImport(\"" + import_ + "\");\n");
+                                usedTypes.add(import_);
                             } else if (afterImports.contains(import_)) {
                                 recipe.append("                    maybeAddImport(\"" + import_ + "\");\n");
                             }
                         }
+                        if (usedTypes.size() == 1) {
+                            preconditions.add("new UsesType<>(\"" + usedTypes.iterator().next() + "\", false)");
+                        } else if (1 < usedTypes.size()) {
+                            preconditions.add("Preconditions.and(" + usedTypes.stream().map(t -> "new UsesType<>(\"" + t + "\", false)").collect(Collectors.joining(", ")) + ')');
```

Comment:
Will this `and()` work correctly in the presence of multiple before templates? If a type is used by all before templates we can `and` it, others will probably have to be `or`ed. So we would have to build up a tree, basically.

Reply:
It should; we build up an `and()` precondition per before template, looking only at imports present in the specific before template. The [further down](https://github.com/openrewrite/rewrite-templating/pull/24/files#diff-51aa8b78b7dec3df0eb18b0aee93bba1f4ea79f3b797ddddaa116cbb0bea5318R360-R364) we compose those together into a `or()` precondition, using each of the before template `and()` precondition.

At both stages I took care to generate the minimal `Precondition` we could get away with, but I started out testing with  a more nested approach by default in 37a2b8c5e8e91e92614d4d2616550f574c1c9134, which is still generated when needed.